### PR TITLE
Support for GitLab projects inside subgroups

### LINKED ISSUE #374: Gitlab subgroups not supported
The URLs for GitLab projects involving subgroups are considered invalid:

Example URL: https://gitlab.com/group/subgroup/project

This makes me impossible to register the package for such projects.

### PR DESCRIPTION
**Pull request by the-Mitu**
This pull requests allows to register packages for GitLab projects placed into subgroups.

Currently the problem is that URL validation throws an error, because it expects url like *service*[.com]//*owner*/*project*. However, GitLab's subgroups allow URLs like gitlab.com/*owner*/*subgroup*/*subsubgroup*/*project*.

This pull request renames **m_project** to **m_projectPath** in **GitLabRepository** and allows it to store the *subgroup*/*subsubgroup*/*project* as is with slashes - so the first part after the URL is put into **m_owner** as it was before, but the rest is joined and put into **m_projectPath** even if contains slashes.

The regex is adjusted as well so that multiple slashed parts are allowed.

This will fix #374.

### CODE REVIEW

**Comment by the-Mitu** [comment]
Unfortunately I have only private projects in subgroups. I've tested it locally on them and it worked but I expect you want to test it as well.

I've browsed the public GitLab groups and found some projects laying inside a subgroup that is publicly available. Example one: https://gitlab.com/ezmaxtech/ha/register

That one is actually not a D project, but maybe it is sufficient to see that the repo is accessed and an error appears just because there is are no dub projects in it. Alternatively a test group/subgroup has to be created.

Thanks for noting the #349 - however I actually need to set up my own instance for hosting non-public packages, so I'll hope I'm more lucky. :)

**Comment by the-Mitu** [comment]
@wilzbach, Tried to test on staging - "Unknown repository type: gitlab". Haven't you forgotten to set the GitLab's server URL in settings.json?

I'd actually suggest it to be "gitlab.com" by default unless another one is specified - not unsupported unless specified - but this is something for another ticket.

**Comment by wilzbach** [comment]
@the-Mitu 
- the staging server is very new (we set it up this week)
- there's no settings.json on Heroku, only environment variables

> this is something for another ticket.

https://github.com/dlang/dub-registry/pull/379
tl;dr: after #379 we should be able to test this here too.

> I've browsed the public GitLab groups and found some projects laying inside a subgroup that is publicly available. Example one: https://gitlab.com/ezmaxtech/ha/register

Hmm I think it needs a valid dub.json / dub.sdl
Anyhow, I created https://gitlab.com/dub-registry-test/a-subgroup/test (it's a fork of https://gitlab.com/dub-registry/test).

> however I actually need to set up my own instance for hosting non-public packages, so I'll hope I'm more lucky. 

We have a #dub channel on the D Slack where you might be able to find help if you aren't lucky.
Simply send me a ping if you want an invite.

**Comment by wilzbach** [comment]
> after #379 we should be able to test this here too.

Merged and thus I rebased this PR. We should be able to test https://gitlab.com/dub-registry-test/a-subgroup/test in a few minutes.

File: views/my_packages.register.dt [hunk 1]
```diff
@@ -15,13 +15,13 @@ block body
...
 			p.light
-				a.blind(href="#{req.rootDir}publish") Learn more 
+				a.blind(href="#{req.rootDir}publish") Learn more
 				| about publishing.
 
 		form(method="POST", action="#{req.rootDir}register_package")
 			p
 				label(for="project") Repository URL
-				input(type="text", name="url", value=url, pattern="^(https?://)?(www\.)?(github(\.com)?|gitlab(\.com)?|bitbucket(\.org)?)/[^/]+/[^/]+([?#.].*)?$", placeholder="github.com/user/repository")
+				input(type="text", name="url", value=url, pattern="^(https?:\/\/)?(www\.)?(github(\.com)?|gitlab(\.com)?|bitbucket(\.org)?)(/[^/]+)+/[^/]+([?#.].*)?$", placeholder="github.com/user/repository")
```
Comment (wilzbach) [inline views/my_packages.register.dt]: This allows invalid patterns for GitHub & Bitbucket, e.g. github.com/foo/bar/barrrr

File: source/dubregistry/dbcontroller.d [hunk 2]
```diff
@@ -509,9 +509,13 @@ struct DbRepository {
...
 			throw new Exception("Invalid Repository URL (missing project)");
-		project = path.front.name;
+
+		if(kind == "gitlab")
+			project = path.map!"a.name".join("/");
+		else
+			project = path.front.name;
 		path.popFront;
-		if (!path.empty)
+		if (!path.empty && kind != "gitlab")
```
Comment (wilzbach) [inline source/dubregistry/dbcontroller.d]: This would allow gitlab.com/foo/bar/foobar/fooobar/foo -  is this really possible?

edit: oh it looks like subgroups can be nested in subgroups

A comment about this in the source code would help future readers...

**Comment by wilzbach** [comment]
>  We should be able to test https://gitlab.com/dub-registry-test/a-subgroup/test in a few minutes.

https://dub-registry-staging-pr-378.herokuapp.com/packages/gitlab-test-package :)

File: views/my_packages.register.dt (hunk 1 above)
Comment (the-Mitu) [reply views/my_packages.register.dt]: Yes, it will. The right regex is:

    ^(https?:\/\/)?(www\.)?(((github(\.com)?|bitbucket(\.org)?)\/[^/]+\/[^/]+([?#.].*)?)|(gitlab(\.com)?(\/[^/]+)+\/[^/]+([?#.].*)?))$

I'll put it in the code tomorrow and commit the fix. In the meantime I would be grateful if anyone tested it once again to confirm its correctness, as it gets long now.

File: source/dubregistry/dbcontroller.d (hunk 2 above)
Comment (the-Mitu) [reply source/dubregistry/dbcontroller.d]: I confirm, GitLab's subgroups are nestable. I'll add the comment as well

Comment (the-Mitu) [reply source/dubregistry/dbcontroller.d]: Comment added.

**Comment by the-Mitu** [comment]
@wilzbach, could I get that invitation to Slack workspace? I'm probably affected by the error you mentioned - and also have a few questions on other topic as well.

**Comment by wilzbach** [comment]
@the-Mitu Invitation is out.
